File: 202403/2958.py
class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        d = defaultdict(int)
        j = 0
        ans = 0
        n = len(nums)
        for i in range(n):
            d[nums[i]] += 1
            while d[nums[i]] > k:
                d[nums[j]] -= 1
                j += 1
            ans = max(ans, i - j + 1)
        return ans

File: 202608/2958.cpp
class Solution {
public:
  int maxSubarrayLength(vector<int> &nums, int k) {
    int ans = 0;
    int j = 0;
    unordered_map<int, int> cnt;
    for (int i = 0; i < nums.size(); ++i) {
      ++cnt[nums[i]];
      while (cnt[nums[i]] > k) {
        --cnt[nums[j++]];
      }
      ans = max(ans, i - j + 1);
    }
    return ans;
  }
};
